Har 3673 indeksiraportti

Kauko opened the pull request:
Alunperin lähdettiin korjaamaan puutetta, missä indeksitarkastusraportti ei varoittanut käyttäjää, jos indeksejä puuttui. Lopulta jouduttiin tekemään aika paljon lisätyötä. Tässä mergessä on poistettu alkuperäinen tapa esittää raporteilla virheitä (eli [:virhe ""], [:info ""] jne), nykyään käytetään [:varillinen-teksti {:arvo ""}] soluja. Samalla excel-raporttimoottorille lisättiin tuki solujen formatoinnille (esim rahaksi), lihavoinnille, korostamiselle.

Vastaisuudessa pitää käydä raportit läpi, ja laittaa kaikki raportit käyttämään mallia, missä itse raportti EI formatoi solujaan lopulliseen muotoon, vaan antaa raporttimoottorille "raa'an datan", ja sarakkeelle tiedon, mitä formatteria kuuluu käyttää.

Kauko committed:
Yhtenäistä raporttien virheenraportointia

Ensiksi raporteille lisättiin tuki korostaa virheitä (esim indeksien puuttuminen)
antamalla solulle arvona vektori, jonka ensimmäisenä arvona on joko
:virhe, :varoitus, tai :info, ja toisena arvona solun varsinainen arvo.
Sittemmin on lisätty :varillinen-teksti, joka tekee käytännössä saman
asian. Tässä commitissa poistetaan ensiksi mainittu versio (koska jälkimmäinen
on commitin hetkellä laajemmassa käytössä). Samalla :varilliselle-tekstille
lisätään :tyyli avain, jota pyritään käyttämään valmiiksi annetun värin sijaan.
Näin eri raporttimoottorit voivat mahdollisesti itse määritellä, mitä esim.
:info :tyyppiselle kentälle halutaan tehdä, ja :varoituksille voidaan tehdä
muutakin, kuin muuttaa väriä.

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
WIP indeksiraportti

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
fmt/pvm-opt palauttaa itse arvon, jos se ei ole pvm

Kauko committed:
Excel-raporttimoottori tukee formaatteja ja värikorostuksia

Solun formaatit (päivämäärä, raha..) kaivetaan maagisilla indekseillä,
URL on kommentoitu relevanttiin paikkaan. Esim päivämäärä olisi
aivan mahdollista formatoida myös itse luotavalla formaatilla
(jota varten on olemassa funktio), jolloin voitaisiin varmistaa,
että solussa päivämäärä on aina suomalaisessa muodossa, mutta
tällöin Excelissä solun kategoriaksi tulee "custom".

Kauko committed:
HTML-raportin raporttielementit tukevat formattereita

Aiemmin :varillinen-teksti tyyppiset solut eivät
kunnioittaneet sarakkeen formaattia, vaan arvoa käytettiin
formatoimatta. Moottoreille pitäis antaa formatoimatonta dataa,
ja tieto formatoijasta, jotta esim excelissä osataan asettaa
solun arvoksi oikea kategoria.

Kauko committed:
PDF-moottorin raporttielementit kunnioittavat :fmt asetusta

Kauko committed:
Lisää raportti-domainiin tarpeellisia apufunktioita

Kauko committed:
Vaihda [:info "teksti"] käyttämään "[:varillista-tekstia {}]"

Kauko committed:
Indeksiraportti näyttää varoituksen jos indeksi puuttuu

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Excel tukee myös korostusta

Samalla palautettiin muodosta-solu multimetodi, sillä
solun luominen vaatii erilaisen metodin tyylin luomisen takia.

Kauko committed:
Opt formatterit käsittelevät myös arvot, jotka eivät ole oikeaa tyyppiä

Esim jos pvm-opt formatterille syöttää arvon, joka ei ole
pvm, palauttaa formatteri itse arvon. Tämä sen takia, että
formatoituun sarakkeeseen voidaan nyt syöttää myös esim virheviestejä.

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Vaihda yrita-funktio raportti-domainiin

Raporteille voidaan antaa sarake, johon käytetään esim prosentti-formatteria.
Kuitenkin itse sarake voi sisältää vaikka varoitustekstin, jolloin
formatteri-funktio epäonnistuu. Aiemmassa commitissa laitoin
-opt formatterit käyttämään yrita-funktiota, mutta katseloimoinnissa
todettiin, että olisi parempi, että formatterit eivät piilota
virheitä, vaan raporttimoottorit käyttävät itse yrita-funktiota.

Kauko committed:
HTML ja PDF raporttimoottorit käyttävät rahan formatointiin euro-formatteria

Kauko committed:
Numeroformatterit heittävät virheen frontilla, eivätkä palauta "epäluku"

Kauko committed:
Excel-raporttimoottori lihavoi yhteenvetorivin

Kauko committed:
Poista turhia riippuvuuksia erilliskustannusraportista

Kauko committed:
Muutkin excelin erikoissolut palauttavat [arvo tyyli] vektorin

Kauko committed:
Typo kelitarkastusraportissa, pvm fmt sarakkeelle

Kauko committed:
Palauta siltatarkastusraportin lihavoinnit ja korostukset

Kauko committed:
Raporttielementteja?-funktio toimii myös riveille, jossa käytetään lisäoptioita

Kauko committed:
Poista siltaraportin sarakkeista turhat :tyyppi avaimet

Kauko committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Excel käsittelee prosentteja samoin kuin muutkin raporttimoottorit

Prosenttiformatoija ottaa käytännössä vastaan luvun, ja laittaa sen
perään prosentin, esim 25 -> 25%. Excel odottaa, ehkäpä loogisesti,
että 0,25 -> 25%, ja 25 -> 2500%. Jos sarakkeen formatter on prosentti,
jaetaan luku sadalla, jotta excel osaa näyttää "oikean" prosenttiluvun.

Kauko committed:
Lisää raporttisolun-arvo testiapuri

Kauko committed:
Korjaa suolasakkoraportin testit

Kauko committed:
Korjaa siltatarkastusraportin testit

Kauko committed:
Siirrä raporttielementin-arvo testiapuri raporttiapureiden namescapeen

Kauko committed:
Korjaa PDF-testit; moottori ei enää heitä niin helposti poikkeusta

Aiemmin muodosta-pdf multimetodi heitti poikkeuksen, jos annettu
solu ei ole ns. raporttielementti. Nyt kyseistä metodia voidaan kutsua
myös soluille, jotka eivät ole raporttielementtejä, vaan puhtaita arvoja.

Multimedoille on myös lisätty :default toteutus, kuten muillakin moottoreilla.
Voi olla perusteltua, että :default metodia ei ole, eli tunnistamattomat
raporttielementit aiheuttavat poikkeuksen.

Kauko committed:
Indeksitarkarkastusraportin testit ottavat huomioon raporttielementit

Kauko committed:
Korjaa ykshinttyot tehtävittäin raportin testit

Kauko commented:
Aiemmin ainakin pdf-moottori heitti poikkeuksen, jos kutsuttiin (muodosta-pdf) metodia asialle, joka ei ole raporttielementti. Kyseinen assert piti poistaa, koska nyt sarakkeet voivat sisältää virheviestejä JA pelkkiä arvoja. Assertin puuttuminen on ikävää, ja sen saisi ehkä pienellä työllä takaisin, mutta en ole varma, olisiko se sen arvoista kuitenkaan.

jarnovayrynen committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

Kauko committed:
Katselmointikommentti: lisää dokumentaatiota

Kauko committed:
Merge branch 'HAR-3673_indeksiraportti' of github.com:finnishtransportagency/harja into HAR-3673_indeksiraportti

tatut committed:
Merge branch 'develop' into HAR-3673_indeksiraportti

tatut committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HAR-3673_indeksiraportti

tatut committed:
Erheen väri ei käytössä

erno commented:
Testaushuomioita yleisesti raporteista, ei välttämättä liity tämän haaran muutoksiin:

Back-nappi voisi palata takaisin generointilomakkeelle.

Sain virheilmoituksen "raportin suoritus epäonnistui" kun painoi tee raportti-nappia jostain tuolla testikoneella mutta en harmillisesti kirjannut sitä heti ylös enkä
löytänyt kohtaa uudelleen kun yritin kirjata sitä tähän. Ehkä backendin logeista löytyisi stack trace?


Sain työmaakokousraportin parametrilomakkeen tilaan jossa aikaväli- ja hoitokausivalinnat
on disabloituna kuten myös raportingenerointinapit. Se saattoi liittyä siihen että kävi ensin
toisessa raporttityypissä jossa on "valittu aikaväli" -checkbox, ja sitten vaihtaa työmaakokousraporttiin.
Ei toistunut suoraviivaisesti, toisela kerralla sain kaikki aika/hoitokausivalinnat disabloitumaan mutta tällä kertaa tee raportti-nappi oli päällä. ohessa ruutukaappaus:

![raportti-valinnat-disabloitu](https://cloud.githubusercontent.com/assets/238642/21226816/e45653a0-c2df-11e6-8cab-18cced495872.png)

Itse generoitujen raporttien sisällöstä en löytänyt mitään ongelmia.

tatut committed:
Käytä ns dokumentaatiota kuolleen koodin sijasta

Keskustelun jälkeen Jari veti ässän hihasta, että ns tason
dokumentaatio, kuten Clojure coressa usein, on hyvä paikka
huomioitaville asioille. Siirrettiin kuollut koodi sinne dokumentaatioksi.

tatut commented on src/clj/harja/palvelin/raportointi/excel.clj:
```diff
@@ -122,20 +130,44 @@
+                                        ;; Tyylejä voi määritellä itse (https://poi.apache.org/apidocs/org/apache/poi/xssf/usermodel/XSSFDataFormat.html)
+                                        ;; tai voimme käyttää valmiita, sisäänrakennettuja tyylejä.
+                                        ;; http://poi.apache.org/apidocs/org/apache/poi/ss/usermodel/BuiltinFormats.html
+                                        :raha (.setDataFormat tyyli 8)
```
Tämä multimetodin alkaa kasvaa pitkäksi. Voisiko tämän casen siirtää ns tasolle mäpiksi fmt->format, jossa olisi esim {:raha 8 :prosentti 10} jne.

tatut committed:
Merge branch 'develop' into HAR-3673_indeksiraportti
